File: elixir/lib/algorithms/quick_sort.ex
defmodule Algorithms.QuickSort do
  @moduledoc """
  Quick sort algorithm with median-of-three pivot selection.
  Time: O(n log n) average, O(n²) worst case
  Space: O(n)
  """

  @doc "Sort a list using quick sort"
  @spec sort(list()) :: list()
  def sort([]), do: []
  def sort([x]), do: [x]

  def sort(list) do
    pivot = median_of_three(list)
    {lesser, equal, greater} = partition(list, pivot)
    sort(lesser) ++ equal ++ sort(greater)
  end

  @doc "Sort with custom comparison function"
  @spec sort_by(list(), (any(), any() -> :lt | :eq | :gt)) :: list()
  def sort_by([], _compare), do: []
  def sort_by([x], _compare), do: [x]

  def sort_by(list, compare) do
    pivot = median_of_three_by(list, compare)
    {lesser, equal, greater} = partition_by(list, pivot, compare)
    sort_by(lesser, compare) ++ equal ++ sort_by(greater, compare)
  end

  defp median_of_three([a]), do: a
  defp median_of_three([a, b]), do: min(a, b)

  defp median_of_three(list) do
    first = hd(list)
    last = List.last(list)
    mid = Enum.at(list, div(length(list), 2))

    [first, mid, last]
    |> Enum.sort()
    |> Enum.at(1)
  end

  defp median_of_three_by([a], _compare), do: a
  defp median_of_three_by([a, _b], _compare), do: a

  defp median_of_three_by(list, compare) do
    first = hd(list)
    last = List.last(list)
    mid = Enum.at(list, div(length(list), 2))

    [first, mid, last]
    |> Enum.sort(fn a, b -> compare.(a, b) != :gt end)
    |> Enum.at(1)
  end

  defp partition(list, pivot) do
    {lesser, rest} = Enum.split_with(list, &(&1 < pivot))
    {equal, greater} = Enum.split_with(rest, &(&1 == pivot))
    {lesser, equal, greater}
  end

  defp partition_by(list, pivot, compare) do
    {lesser, rest} = Enum.split_with(list, &(compare.(&1, pivot) == :lt))
    {equal, greater} = Enum.split_with(rest, &(compare.(&1, pivot) == :eq))
    {lesser, equal, greater}
  end
end

File: haskell/src/Algorithms/QuickSort.hs
{-|
Module      : Algorithms.QuickSort
Description : Quick sort algorithm
Stability   : stable

A purely functional quick sort implementation with median-of-three pivot.
Time: O(n log n) average, O(n²) worst case
Space: O(n) - creates new lists
-}
module Algorithms.QuickSort
  ( sort
  , sortBy
  ) where

-- | Sort a list using quick sort
sort :: Ord a => [a] -> [a]
sort = sortBy compare

-- | Sort with custom comparison function
sortBy :: (a -> a -> Ordering) -> [a] -> [a]
sortBy _ [] = []
sortBy _ [x] = [x]
sortBy cmp xs = sortBy cmp lesser ++ [pivot] ++ sortBy cmp greater
  where
    pivot = medianOfThree cmp xs
    lesser  = filter (\x -> cmp x pivot == LT) (filter (/= pivot) xs)
    greater = filter (\x -> cmp x pivot /= LT) (filter (/= pivot) xs)

-- | Select median of first, middle, and last elements
medianOfThree :: (a -> a -> Ordering) -> [a] -> a
medianOfThree _ [x] = x
medianOfThree _ [x, _] = x
medianOfThree cmp xs = median3 cmp first mid lastElem
  where
    first = head xs
    lastElem = last xs
    mid = xs !! (length xs `div` 2)

-- | Find median of three values
median3 :: (a -> a -> Ordering) -> a -> a -> a -> a
median3 cmp a b c
  | (cmp a b /= GT && cmp b c /= GT) || (cmp c b /= GT && cmp b a /= GT) = b
  | (cmp b a /= GT && cmp a c /= GT) || (cmp c a /= GT && cmp a b /= GT) = a
  | otherwise = c
